server/src/main/java/org/opensearch/tasks/Task.java

```diff
@@ -78,6 +80,17 @@ public class Task {
 
     private final Map<Long, List<ThreadResourceInfo>> resourceStats;
 
+    private final List<TaskResourceTrackingListener> resourceTrackingListeners;
+
+    private final AtomicBoolean isResourceTrackingCompleted = new AtomicBoolean(false);
```

Comment:
As start/stop of thread execution is not guaranteed to happen before task unregistration, it may be possible that a thread starts and stops after the resource tracking is already completed. If we just rely on the counter, it would transition from `0 -> 1 -> 0` which would invoke the callback more than once.

Reply:
If its just about listeners can you please use `NotifyOnceListener`

Reply:
Thanks for the suggestion!